x/wasm/IBC.md

```diff
@@ -0,0 +1,253 @@
+func (c *mockContract) IBCRecvPacket(ctx sdk.Context, k *wasm.Keeper, env IBCInfo, msg []byte) (response []byte, err error) {}
+
+// how to handle error here? if we push it up the ibc stack and fail the transaction (normal handling),
+// the packet may be posted again and again. just log and ignore failures here? what does a failure even mean?
+func (c *mockContract) IBCPacketAck(ctx sdk.Context, k *wasm.Keeper, env IBCInfo, originalMsg []byte, result []byte) (response []byte, err error) {}
+
+// same question as for IBCPacketAck
+func (c *mockContract) IBCPacketFailed(ctx sdk.Context, k *wasm.Keeper, env IBCInfo, originalMsg []byte, errorMsg string) (response []byte, err error) {}
+
+type IBCInfo struct {
+	// PortID of the remote contract???
+	RemotePortID string
+	// PortID of the our contract???
+	OurPortID string
+	// ChannelID packet was sent on
+	ChannelID string
+	Packet    *IBCPacketInfo `json:"packet,omitempty"`
+}
+
+// do we need/want all this info?
+type IBCPacketInfo struct {
+	Sequence uint64
+	// identifies the port on the sending chain.
+	SourcePort string
+	// identifies the channel end on the sending chain.
+	SourceChannel string
+	// block height after which the packet times out
+	TimeoutHeight uint64
+	// block timestamp (in nanoseconds) after which the packet times out
+	TimeoutTimestamp uint64
+}
+```
+
+Channel Lifecycle:
+
+**TODO**
+
+Queries:
+
+**TODO**
+
+### Contract (Wasm) entrypoints
+
+**TODO**
+
+## Ports and Channels Discussion
+
+We decided on "one port per contract", especially after the IBC team raised
+the max length on port names to allow `wasm-<bech32 address>` to be a valid port.
+Here are the arguments for "one port for x/wasm" vs "one port per contract"
+
+### One Port per Contract
+
+This may be the most straight-forward mapping, treating each contract 
+like a module. It does lead to very long portIDs however. Make special
+attention to both the Channel establishment (which should be compatible
+with standard ICS20 modules without changes on their part), as well
+as how contracts can properly identify their counterparty.
+
+* Upon `Instantiate`, if a contract is *IBC Enabled*, we dynamically 
+  bind a port for this contract. The port name is `wasm-<contract address>`,
+  eg. `wasm-cosmos1hmdudppzceg27qsuq707tjg8rkgj7g5hnvnw29`
+* If a *Channel* is being established with a registered `wasm-xyz` port,
+  the `x/wasm.Keeper` will handle this and call into the appropriate
+  contract to determine supported protocol versions during the
+  [`ChanOpenTry` and `ChanOpenAck` phases](https://docs.cosmos.network/master/ibc/overview.html#channels).
+  (See [Channel Handshake Version Negotiation](https://docs.cosmos.network/master/ibc/custom.html#channel-handshake-version-negotiation))
+* Both the *Port* and the *Channel* are fully owned by one contract.
+* Question: `x/wasm` only accepts *ORDERED Channels* for simplicity of contract
+  correctness? Or should we allow the contract to decide?
```

Comment:
IMHO there is no need to limit this in wasmd. Let the contract (framework) decide so that we have the most flexible solution in wasmd for future use cases.

Reply:
Okay, we allow both ordered and unordered. And recommend ordered unless the developer can reason about distributed systems well.